### DISCUSSION TITLE
State-dependent actions and gen function

### DISCUSSION THREAD

**Discussion by Cbroomer** [JuliaPOMDP/POMDPs.jl#382] [Q&A] [answered]
Hello,

I am new in this field and I would like you to ask you 2 questions that right now are very important for me:

Is it possible to use state-dependent actions with these solvers (https://github.com/JuliaPOMDP/TabularTDLearning.jl) ? Maybe a function that given a state returns a vector with all the possible actions?

How does the generative model really works? I mean if I use gen = function (s, a, rnd) it takes the initial state and then tries all the possible actions of the actions space according to their keys? I want to know if actions are selected in order or are chosen randomly.

I hope you can help me, thank you so much.

Best regards.

**Comment by lassepe** [comment, marked as answer]
Hi @clrescobar,

Unfortunately, `TabularTDLearning` does not currently support state-dependent actions since that would make the table "non square". That said, it would be possible to make that extension by using another data structure for the Q-values in `TabularTDLearning`. That, however, would require minor changes to the solver/code over there.

The [`gen`](https://juliapomdp.github.io/POMDPs.jl/latest/def_pomdp/#Using-a-single-generative-function-instead-of-separate-T,-Z,-and-R) function implements the (PO)MDP transition (and observation) model in a generative representation. That is, rather than having to provide probability densities over states, observations, and rewards via `T`, `Z`, and `R`, you just have to implement `gen` as a function to *sample* from the joint distribution of the three. Therefore, when you call `gen(s, a, rng)`, you will get a random sample `(;sp, o, r)` of a new state `sp`, observation `o`, and reward `r`  conditioned on your previous state `s` and action `a`. If your problem is non-deterministic and you repeatedly call this function, repeated calls to this function may yield different `(; sp, o, r)` tuples. However, the distribution of those samples will be consistent with the POMDP model (by definition, has the `gen` function implicitly *defines* this distribution).

A quick side note: In practice you should never have to call `gen` yourself. Instead, you would use the [`@gen`](https://juliapomdp.github.io/POMDPs.jl/latest/api/#POMDPs.@gen) macro which combines generative and explicit models into a single sampling call.

**Reply by Cbroomer** [reply]
Hello @lassepe 

First of all thank you so much for your quick response.

Regarding Q-tables, is it possible to acces to them in order to see them? It will be really usefull.

On the other side, with respect to the `gen` function, the order in which actions are called is random? I mean, given an initial state `s`, in order to know the next state `sp` and reward `r`, which action is selected? One random action from the action-space or actions are selected following the order of their keys?

Thank you so much!

**Reply by lassepe** [reply]
You should be able to access the q values via the solver. E.g. for SARSA:
https://github.com/JuliaPOMDP/TabularTDLearning.jl/blob/3630ddfdb1bac2e95912fa9648e14d6fb6831563/src/sarsa.jl#L33

Regarding `gen`: the action `a` is input to the function so `gen` does not decide in which order actions are used. `gen` is used to simulate transitions of the model. This kind of simulation is used for evaluation or inside a solver to simulate the outcome of an action.

**Reply by Cbroomer** [reply]
Hello,

Thanks! Which command should I use to access the q-table?

So, how is the order in which actions are called? Because in my case I am using state-dependent actions (using a dummy state) so I guess the long times of computations are because of the order in which actions are called.

**Reply by lassepe** [reply]
All the methods in TabularTDLearning.jl take an `exploration_policy` upon construction and that policy is used to select the actions during learning:

https://github.com/JuliaPOMDP/TabularTDLearning.jl/blob/3630ddfdb1bac2e95912fa9648e14d6fb6831563/src/q_learn.jl#L56

The object that you get back is a [`ValuePolicy`](https://github.com/JuliaPOMDP/POMDPPolicies.jl/blob/6249800d9b21d5f322e8505caa14a8b331e7ad39/src/vector.jl#L53) which has a property `value_table` which you can access to get the q-values.

In general, the code for tabular q-learning is very short. Those 92 lines of code [here](https://github.com/JuliaPOMDP/TabularTDLearning.jl/blob/3630ddfdb1bac2e95912fa9648e14d6fb6831563/src/q_learn.jl) are all there is to it. Maybe some of the questions that you have about the solver internals may also be answered by taking a look at that. It's pretty close to text book pseudo code.

**Reply by Cbroomer** [reply]
Perfect, thank you so much for your help. Which solvers can work with state-dependent actions?

**Reply by lassepe** [reply]
I have no good overview which solvers allow this. I guess @zsunberg may know? If so, maybe we could add this information to an overview table to the Docs or the README.

**Reply by lassepe** [reply]
FWIW, [MCTS.jl](https://github.com/JuliaPOMDP/MCTS.jl) certainly supports this but of course that is more on the "planning" than on the "learning" end of solvers.

**Reply by Cbroomer** [reply]
No problem, regarding Q-Table is it possible to assign values to it? So for example, if there are impossible actions we can previously write on the Q-Table a bad value since the beginning?

**Reply by lassepe** [reply]
Yes, have a look [here](https://github.com/JuliaPOMDP/TabularTDLearning.jl/blob/3630ddfdb1bac2e95912fa9648e14d6fb6831563/src/q_learn.jl#L45). If the `solver` as some `solver.Q_vals` set upon the call to `POMDPs.solve(solver, mdp)` it will copy those over to the initial guess of the q values. So you could set illegal actions to infinite cost and then specify an exploration policy which takes actions of that cost with zero probability.

**Reply by Cbroomer** [reply]
Great, thanks! Maybe a softmax policy better than epsilon greedy?

**Reply by lassepe** [reply]
I don't think softmax works if you set Q-values to Inf because `Inf / Inf == NaN`. You would then just have to set the Q-values to a "sufficiently high but non inf value". But a better way would perhaps be to implement a custom epislon-greedy variant which prunes those actions which have infinite q-value.You can use the [CategoricalTabularPolicy](https://github.com/JuliaPOMDP/POMDPPolicies.jl/blob/170414e2ea775f1f4ccfc6617d2c00388659cdbf/src/stochastic.jl#L48) as a starting point for your implementation.

**Reply by lassepe** [reply]
The implementation is left as an exercise to the reader ;)

**Reply by Cbroomer** [reply]
So, before using the solver values have to be set in the q-table right?

**Reply by zsunberg** [reply]
@lassepe Thanks a bunch for your answers and for pointing to good places in the source code.

> So, before using the solver values have to be set in the q-table right?

Yes, it would be something like
```julia
solver.Q_vals = zeros(length(states(mdp)), length(actions(mdp)))
solve(solver, mdp)
```
However, a better approach would probably be to penalize the actions themselves in the mdp definition with something like the following if you are implementing the `gen` function instead of `transition` and `reward`:
```julia
gen = function(s, a, rng)
    if invalid(s, a)
        return (sp = s, r = -1e9)
    end
    # the rest of your dynamics
end
```

> Which solvers can work with state-dependent actions?

Are you trying to solve an MDP or a POMDP? For MDPs, MCTS and Value Iteration work with state-dependent actions. (Note: If you can define an explicit transition distribution, value iteration will get the optimal value much faster than the Q-learning or SARSA methods)

If you want to use Q learning or SARSA, it would not be hard to modify the source code to handle the state-dependent actions. As @lassepe has been suggesting, in Julia it is often fairly straightforward to look at the source and see what is going on. Then you can modify it to do what you want. Let us know if you try this and run into problems.

**Reply by Cbroomer** [reply]
Hello,

I am trying to solve and MDP  implementing the ``gen`` function. In fact, I am penalizing non possible the actions in the MDP, but instead of ``sp=s`` I have created a dummy state following [this](https://juliapomdp.github.io/POMDPs.jl/latest/faq/#How-can-I-implement-terminal-actions? ) which is terminal. Is it better ``sp=s``? Thanks :)

**Reply by zsunberg** [reply]
A terminal state is actually better!

**Reply by Cbroomer** [reply]
I have tried both and when I use ``sp=s`` is faster than when I use the dummy state. I do not understand why...It makes no sense.

**Reply by zsunberg** [reply]
Is the terminal state a different type than the other states? If the compiler always knows that the state is going to be a certain type, it can generate much faster code - for instance see the Julia Performance Tips: https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-abstract-container

**Reply by Cbroomer** [reply]
There are terminal states and the dummy state which is also terminal.  So I concatenate terminal and dummy states as terminal and then used termination:
```
terminal = vcat(terminal,spt)
termination(s::State) =  s in terminal

.....
    states = SS, #State-space
    actions = AS, #Action-space
    initialstate = init_resupply, #Initial state
    discount = gamma, #Discount factor
    isterminal = termination) #Termination: Dummy or terminal state

````

**Reply by zsunberg** [reply]
It's hard to know the exact source of the performance difference from just a small snippet of code, but it may be because the `sp=s` version of the `gen` function is "type stable" while the `sp=spt` version is not (see https://docs.julialang.org/en/v1/manual/performance-tips/#Write-%22type-stable%22-functions). Additionally, if `spt` is a global variable, it might cause extra slow performance: https://docs.julialang.org/en/v1/manual/performance-tips/#Avoid-global-variables. In general, skimming the "Performance Tips" section of the Julia manual is really helpful in understanding why code is faster or slower.

**Reply by Cbroomer** [reply]
I mean faster in number of iterations, I created ``spt`` from ``struct State`` and ``s`` is also ``State``, so I guess this is because ``spt`` is a global variable.
Thank you so much for your help.

**Reply by zsunberg** [reply]
> I mean faster in number of iterations,

Oh, that is more interesting! It makes sense though: When the agent tries a prohibited action with `sp=spt`, it has to start a completely new episode. With `sp=s`, it gets to keep going on the same episode. It's like having to start over completely on a video game when the character dies vs having multiple lives - things are much easier with multiple lives.

**Reply by Cbroomer** [reply]
Perfect, thank you so much for all you help!
